fdb-record-layer-core/src/test/java/com/apple/foundationdb/record/provider/foundationdb/MockedLocalityUtil.java

```diff
@@ -0,0 +1,184 @@
+     *
+     * <p>
+     * This is method randomly partitions the sorted list of {@code keys} among {@code rangeCount} servers as if the keys
+     * were stored contiguously on those servers. The outcome of this partition is later used by {@link #getBoundaryKeys(Transaction, byte[], byte[])}.
+     * Note that {@code getBoundaryKeys} must be called after this method is called, otherwise it will assume no keys exists.
+     * </p>
+     *
+     * @param keys a sorted list of keys
+     * @param rangeCount the number of ranges to return.
+     */
+    @Nonnull
+    public static void init(@Nonnull List<byte[]> keys, int rangeCount) {
+        INSTANCE.ranges = new ArrayList<>(rangeCount);
+        INSTANCE.keyRanges = new ArrayList<>(keys.size());
+
+        // Build the boundaries randomly
+        Random randomGenerator = new Random();
+        int lastBoundaryIndex = 0;
+        for (int i = 0; i < rangeCount; i++) {
+            if (i > 0) {
+                lastBoundaryIndex += randomGenerator.nextInt((keys.size() - INSTANCE.ranges.get(i - 1)) / (rangeCount - i));
+            }
+            INSTANCE.ranges.add(lastBoundaryIndex);
+        }
+
+        // Create a map between keys and their ranges.
+        int rangeIndex = 0;
+        for (int i = 0; INSTANCE.ranges.size() > 0 && i < keys.size(); i++) {
+            while (rangeIndex < INSTANCE.ranges.size() && INSTANCE.ranges.get(rangeIndex) < i) {
+                rangeIndex++;
+            }
+            INSTANCE.keyRanges.add(new KeyRange(keys.get(i), Tuple.fromBytes(keys.get(i)),
+                    rangeIndex == INSTANCE.ranges.size() ? INSTANCE.ranges.get(rangeIndex - 1) /* it's the last range */ : INSTANCE.ranges.get(rangeIndex)));
+        }
+    }
+
+    /**
+     * Return a {@code CloseableAsyncIterator} of keys {@code k} such that {@code begin <= k < end} and {@code k}
+     * represents a location at the start of a contiguous range stored on a single server. Note that this class
+     * fakes the actual locations, and the returned keys are derived from the random boundaries created in {@link #init(List, int)}.
+     *
+     * @param tr the transaction on which to base the query
+     * @param begin the inclusive start of the range
+     * @param end the exclusive end of the range
+     *
+     * @return a sequence of keys denoting the start of the ranges
+     */
+    @Nonnull
+    public CloseableAsyncIterator<byte[]> getBoundaryKeys(@Nonnull Transaction tr, @Nonnull byte[] begin, @Nonnull byte[] end) {
+        // This mocked locality API does not need tr.
+        return new MockedBoundaryIterator(keyRanges, ranges, begin, end);
+    }
+
+    static class MockedBoundaryIterator implements CloseableAsyncIterator<byte[]> {
+        private List<byte[]> ranges;
+        int lastBeginIndex;
+
+        MockedBoundaryIterator(@Nonnull List<KeyRange> keyRanges, @Nonnull List<Integer> ranges, @Nonnull byte[] begin, @Nonnull byte[] end) {
```

Comment:
There should be JavaDoc explaining `keyRanges` and `ranges`

Reply:
Sure, I can add it, but the scope of this and the two fields is not `public`.